pkg/R/DataFrame.R

```diff
@@ -52,21 +53,108 @@ dataFrame <- function(sdf) {
 #' printSchema(df)
 #'}
 
-setGeneric("printSchema", function(df) { standardGeneric("printSchema") })
+setGeneric("printSchema", function(x) { standardGeneric("printSchema") })
 
 setMethod("printSchema",
-          signature(df = "DataFrame"),
-          function(df) {
-            sdf <- df@sdf
-            schemaString <- callJMethod(sdf, "printSchema")
+          signature(x = "DataFrame"),
+          function(x) {
+            schemaString <- callJMethod(x@sdf, "printSchema")
             cat(schemaString)
           })
 
+#' Get schema object
+#' 
+#' Returns the schema of this DataFrame as a list of named lists. Each named
+#' list contains the elements of the StructField type, i.e. name, dataType, nullable.
+#' 
+#' @param x A SparkSQL DataFrame
+#' 
+#' @rdname schema
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' dfSchema <- schema(df)
+#'}
+
+setGeneric("schema", function(x) { standardGeneric("schema") })
+
+setMethod("schema",
+          signature(x = "DataFrame"),
+          function(x) {
+            schemaOut <- callJMethod(x@sdf, "schema")
+            schemaOut
+          })
+
+#' DataTypes
+#' 
+#' Return all column names and their data types as a list
+#' 
+#' @param x A SparkSQL DataFrame
+#' 
+#' @rdname dtypes
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' dtypes(df)
+#'}
+
+setGeneric("dtypes", function(x) {standardGeneric("dtypes") })
+
+setMethod("dtypes",
+          signature(x = "DataFrame"),
+          function(x) {
+              lapply(schema(x), function(f) {
+                field <- c(f$name, f$dataType)
+            })
+          })
+
+#' Column names
+#' 
+#' Return all column names as a list
+#' 
+#' @param x A SparkSQL DataFrame
+#' 
+#' @rdname columns
+#' @export
+#' @examples
+#'\dontrun{
+#' sc <- sparkR.init()
+#' sqlCtx <- sparkRSQL.init(sc)
+#' path <- "path/to/file.json"
+#' df <- jsonFile(sqlCtx, path)
+#' columns(df)
+#'}
+setGeneric("columns", function(x) {standardGeneric("columns") })
+
+setMethod("columns",
+          signature(x = "DataFrame"),
+          function(x) {
+            sapply(schema(x), function(f) {
```

Comment:
Good question. It's because I was mirroring the PySpark API version of `columns`, which grabs it out of the schema (as I'm sure you probably know :P )

The other benefit is that is doing it this way allows us to use `sapply` and return a `character` vector instead of a list. This allow us to still access all the elements inside the result by index location, but vectors print out more cleanly in the console than lists do (at least when all you've got is a series of strings).